libpromises/evalfunction.c

```diff
@@ -3152,6 +3153,69 @@ static FnCallResult FnCallFileStatDetails(ARG_UNUSED EvalContext *ctx,
     {
         return FnFailure();
     }
+    else if (!strcmp(detail, "xattr"))
+    {
+#if defined(WITH_XATTR)
+        // Extended attributes include both POSIX ACLs and SELinux contexts.
+        char attr_raw_names[CF_BUFSIZE];
+        ssize_t attr_raw_names_size = listxattr(path, attr_raw_names, sizeof(attr_raw_names));
+
+        if (attr_raw_names_size < 0)
+        {
+            if (errno != ENOTSUP && errno != ENODATA)
+            {
+                Log(LOG_LEVEL_INFO, "Can't read extended attributes of '%s'. (listxattr: %s)",
+                    path, GetErrorStr());
+            }
+        }
+        else
+        {
+            Buffer *printattr = BufferNew();
+            for (int pos = 0; pos < attr_raw_names_size;)
+            {
+                const char *current = attr_raw_names + pos;
+                pos += strlen(current) + 1;
+
+                if (!StringIsPrintable(current))
+                {
+                    Log(LOG_LEVEL_INFO, "Skipping extended attribute of '%s', it has a non-printable name: '%s'",
+                        path, current);
+                    continue;
+                }
+
+                char data[CF_BUFSIZE];
+                int datasize = getxattr(path, current, data, sizeof(data));
+                if (datasize < 0)
+                {
+                    if (errno == ENOTSUP)
+                    {
+                        continue;
+                    }
+                    else
+                    {
+                        Log(LOG_LEVEL_INFO, "Can't read extended attribute '%s' of '%s'. (getxattr: %s)",
+                            path, current, GetErrorStr());
+                    }
+                }
+                else
+                {
+                  if (!StringIsPrintable(data))
+                  {
+                      Log(LOG_LEVEL_INFO, "Skipping extended attribute of '%s', it has non-printable data: '%s=%s'",
+                          path, current, data);
+                      continue;
+                  }
+
+                  BufferPrintf(printattr, "%s=%s", current, data);
```

Comment:
Oh, right.  I had that originally and forgot to add it back after a minor refactoring!